packages/react-form-renderer/src/files/use-field-api.js

```diff
@@ -1,38 +1,107 @@
+import assignSpecialType from '../form-renderer/assign-special-type';
+import componentTypes from './component-types';
+import { prepareArrayValidator, getValidate } from '../form-renderer/validator-helpers';
+import composeValidators from './compose-validators';
+import isEqual from 'lodash/isEqual';
 
-const useFieldApi = ({ name, initializeOnMount, component, render, validate, dataType, ...props }) => {
-  const { actionMapper } = useContext(RendererContext);
-  const formOptions = useFormApi();
-  const fieldProps = useField(name, { validate, ...props });
+const calculateInitialValue = (props) => {
+  if (Object.prototype.hasOwnProperty.call(props, 'initialValue') && Object.prototype.hasOwnProperty.call(props, 'dataType')) {
+    return convertInitialValue(props.initialValue, props.dataType);
+  }
+};
+
+const calculateArrayValidator = (props, validate, component, validatorMapper) => {
+  if ((validate || props.dataType) && componentTypes.FIELD_ARRAY === component) {
+    return prepareArrayValidator(getValidate(validate, props.dataType, validatorMapper));
+  }
+};
+
+const calculateValidate = (props, validate, component, validatorMapper) => {
+  if ((validate || props.dataType) && componentTypes.FIELD_ARRAY !== component) {
+    return composeValidators(getValidate(validate, props.dataType, validatorMapper));
+  }
+};
+
+const useFieldApi = ({ name, initializeOnMount, component, render, validate, ...props }) => {
+  const { actionMapper, validatorMapper, formOptions } = useContext(RendererContext);
+  const [initialValue, setInitialValue] = useState(() => calculateInitialValue(props));
+  const [arrayValidator, setArrayValidator] = useState(() => calculateArrayValidator(props, validate, component, validatorMapper));
+  const [stateValidate, setValidate] = useState(() => calculateValidate(props, validate, component, validatorMapper));
+  const [type, setType] = useState(() => assignSpecialType(component));
+  const mounted = useRef(false);
+
+  const enhancedProps = {
+    type,
+    ...props,
+    ...(initialValue ? { initialValue } : {}),
+    ...(stateValidate ? { validate: stateValidate } : {})
+  };
+
+  const fieldProps = useField(name, enhancedProps);
+
+  /** Reinitilize type */
+  useEffect(() => {
+    if (mounted.current) {
+      const specialType = assignSpecialType(component);
+      if (specialType !== type) {
+        setType(specialType);
+      }
+    }
+  }, [component]);
+
+  /** Reinitilize array validator/validate */
+  useEffect(() => {
+    if (mounted.current) {
+      setArrayValidator(calculateArrayValidator(props, validate, component, validatorMapper));
+      setValidate(calculateValidate(props, validate, component, validatorMapper));
+    }
+  }, [validate, component, props.dataType]);
+
+  /** Re-convert initialValue when changed */
+  useEffect(() => {
+    if (mounted.current) {
+      const newInitialValue = calculateInitialValue(props);
```

Comment:
we should probably count this value only if the condition for the state update is met. We can save some resources.

Reply:
Not sure if we can make it more conditional. Right now, it's being recalculated when initial value or datatype are changed. Datatype's change should definitely recalculate the value... initialValue only if the value is not of the it's type, so we could add some check function (and also added it to the first calculation)